onnxruntime/core/framework/utils.cc

```diff
@@ -282,7 +282,12 @@ static common::Status SetupFetchesForExecute(const SessionState& session_state,
         }
 
         auto tensor_provider_type = tensor_provider->Type();
-        if (node_provider_type == tensor_provider_type) {
+        if (node_provider_type == tensor_provider_type ||
+            // special case to avoid output copy for nuphar provider
+            // need this to pass test Scan8.UnknownDimInSubgraphOutput in Nuphar
+            // TODO: have a more standard handling for tensor zero-copy between devices
+            (tensor_provider_type == kCpuExecutionProvider && node_provider_type == kNupharExecutionProvider) ||
+            (tensor_provider_type == kNupharExecutionProvider && node_provider_type == kCpuExecutionProvider)) {
           new_fetches[idx] = fetches[idx];
```

Comment:
Hopefully this special case will go away after you merge in Ke Zhang's changes from last week

Reply:
You are right, this is no longer needed.

---
In reply to: [313179497](https://github.com/microsoft/onnxruntime/pull/881#discussion_r313179497) [](ancestors = 313179497)